### PR TITLE
Issue 655 export ir analysis sql

### PR DESCRIPTION
**Pull request by ssuvorov-fls**
fixes https://github.com/OHDSI/WebAPI/issues/655
see also https://github.com/OHDSI/WebAPI/pull/947

### CODE REVIEW

**Comment by chrisknoll** [comment]
Merged from master to resolve conflicts.  please Pull latest.

It looks good to go, except we should add a 'copy to clipboard' for the SQL. can you add that?  There are examples of the 'copy to clipboard' elsewhere in the application, let me know if you need me to find it.

File: js/pages/incidence-rates/ir-manager.html [hunk 1]
```diff
@@ -54,6 +56,89 @@
...
+			<div role="tabpanel" data-bind="css: { active: $component.activeTab() == 'export' }" class="tab-pane">
+				<ul class="nav nav-pills">
+					<li role="presentation" data-bind="css: { active: $component.exportTabMode() == 'sql' }, click: function() { $component.activeTab('sql');}">
+						<a>SQL</a>
+					</li>
+				</ul>
+
+				<div class="tab-content">
+					<div role="tabpanel" data-bind="css: { active: $component.exportTabMode() == 'sql'}" class="tab-pane">
+						<div class="row">
```
Comment (pavgra) [inline js/pages/incidence-rates/ir-manager.html]: This is copy-paste from Cohort Definitions, isn't it? Let's move this to a shared component and not introduce code duplication

File: js/pages/incidence-rates/ir-manager.js [hunk 2]
```diff
@@ -153,6 +164,23 @@ define([
 			});
 			this.expressionMode = ko.observable('import');
 
+			this.generatedSql = {};
+			this.generatedSql.mssql = ko.observable('');
```
Comment (pavgra) [inline js/pages/incidence-rates/ir-manager.js]: Looks like code duplication. Why cannot we do something like:
```
// Place into some app-wide const.js
const dialects = ['mssql', 'oracle', 'postgresql', ...]

// Constructor
dialects.forEach(d => {
  this.generatedSql[d] = ko.observable('');
})

// showSql
const promises = dialects.map(d => {
  return IRAnalysisService.translateSql(result.templateSql, d).then(({data}) => this.generatedSql[d](data.targetSQL));
})
Promise.all(promises).then.....
```

File: js/pages/incidence-rates/tabs/export/bigquery.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+	'components/Component',
+	'utils/AutoBind',
+	'utils/CommonUtils',
+], function (
+	view,
+	Component,
+	AutoBind,
+	commonUtils,
+) {
+	class IRexportbigqueryTab extends AutoBind(Component) {
```
Comment (pavgra) [inline js/pages/incidence-rates/tabs/export/bigquery.js]: Why do you need separate components for each dialect? What's difference between e.g. `IRexportbigqueryTab` and `IRexportimpalaTab`? Why cannot a single component be used?

File: js/pages/incidence-rates/tabs/ir-utilities.html [hunk 4]
```diff
@@ -0,0 +1,19 @@
+<div class="paddedWrapper">
+    <ul class="nav nav-pills">
+        <li role="presentation" data-bind="css: {active: parent.expressionMode() == 'import' }, click: function() {parent.expressionMode('import') };"><a>Import</a></li>
+        <li role="presentation" data-bind="css: {active: parent.expressionMode() == 'export' }, click: function() {parent.expressionMode('export') };"><a>Export</a></li>
+    </ul>
+    <div role="tabpanel" data-bind="visible: parent.expressionMode() == 'import'" class="tab-pane">
+        <div class="paddedWrapper">
+            <textarea class="code" style="width: 100%; height: 300px" data-bind="textInput: parent.importJSON"></textarea>
```
Comment (pavgra) [inline js/pages/incidence-rates/tabs/ir-utilities.html]: No inline CSS please

File: js/services/IRAnalysis.js [hunk 5]
```diff
@@ -148,7 +148,33 @@ define(function (require, exports) {
 				authApi.handleAccessDenied(response);
 				return response;
 			});
-	}	
+	}
+
+	function getSql(expression, analysisId, options) {
+		var getSqlPromise = $.ajax({
```
Comment (pavgra) [inline js/services/IRAnalysis.js]: Use `httpService` please

File: js/pages/incidence-rates/ir-manager.js (hunk 2 above)
Comment (ssuvorov-fls) [reply js/pages/incidence-rates/ir-manager.js]: done

File: js/pages/incidence-rates/tabs/export/bigquery.js (hunk 3 above)
Comment (ssuvorov-fls) [reply js/pages/incidence-rates/tabs/export/bigquery.js]: done

File: js/pages/incidence-rates/tabs/ir-utilities.html (hunk 4 above)
Comment (ssuvorov-fls) [reply js/pages/incidence-rates/tabs/ir-utilities.html]: done

File: js/services/IRAnalysis.js (hunk 5 above)
Comment (ssuvorov-fls) [reply js/services/IRAnalysis.js]: done

File: js/pages/cohort-definitions/cohort-definition-manager.js [hunk 6]
```diff
@@ -806,65 +807,27 @@ define(['jquery', 'knockout', 'text!./cohort-definition-manager.html',
...
+						return this.service.translateSql(sql, p.dialect.value).then(({data}) => p.translate(data.targetSQL));
+					  })
+	
+					Promise.all(promises).then(() => {
+						this.isLoadingSql(false);
 					});
 				});
 			}
 
+			getGeneratedSql(dialect) {
```
Comment (wivern) [inline js/pages/cohort-definitions/cohort-definition-manager.js]: Where it's used?

File: js/pages/incidence-rates/ir-manager.js [hunk 7]
```diff
@@ -111,6 +119,8 @@ define([
 			});
 
 			this.isRunning = ko.observable(false);
+      this.exportTabMode = ko.observable('sql');
```
Comment (wivern) [inline js/pages/incidence-rates/ir-manager.js]: code style (tabs vs spaces)

File: js/pages/incidence-rates/ir-manager.js [hunk 8]
```diff
@@ -176,6 +187,17 @@ define([
 				createdByUsernameGetter: () => this.selectedAnalysis() && this.selectedAnalysis().createdBy()
 			});
 
+			this.translateParams = [];
+			globalConstants.dialects.forEach(d => {
```
Comment (wivern) [inline js/pages/incidence-rates/ir-manager.js]: Use `Array.prototype.map`

File: js/pages/incidence-rates/ir-manager.js [hunk 9]
```diff
@@ -466,6 +488,33 @@ define([
...
+					return IRAnalysisService.translateSql(sql, p.dialect.value).then(({data}) => p.translate(data.targetSQL));
+				  })
+
+				Promise.all(promises).then(() => {
+					this.isLoadingSql(false);
+				});
+			});
+		}
+
+		getGeneratedSql(dialect) {
```
Comment (wivern) [inline js/pages/incidence-rates/ir-manager.js]: Where it's used?

File: js/pages/incidence-rates/tabs/ir-conceptsets.js [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+	Component,
+	AutoBind,
+	commonUtils,
+) {
+	class IRConceptsetsTab extends AutoBind(Component) {
+		constructor(params) {
+      super(params);
+      this.parent = params.parent;
+    }
+	}
```
Comment (wivern) [inline js/pages/incidence-rates/tabs/ir-conceptsets.js]: tabs vs spaces

File: js/pages/incidence-rates/tabs/ir-definition.js [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+	Component,
+	AutoBind,
+	commonUtils,
+) {
+	class IRDefinitionTab extends AutoBind(Component) {
+		constructor(params) {
+      super(params);
+      this.parent = params.parent;
+    }
+	}
```
Comment (wivern) [inline js/pages/incidence-rates/tabs/ir-definition.js]: tabs vs spaces

File: js/pages/incidence-rates/tabs/ir-export.html [hunk 12]
```diff
@@ -0,0 +1,57 @@
+<div data-bind="css: classes()">
+  <div data-bind="css: classes('export-btn')">
+    <span id="copyCopyIRSQLMessage" style="display:none" class="alert alert-success" role="alert"><i class="fa fa-check-square-o" aria-hidden="true"></i><strong>Copied To Clipboard!</strong>
+    </span>&nbsp;&nbsp;<button class="btn btn-sm btn-primary" itle="Copy to clipboard" id="btnCopyIRSQLClipboard" data-bind="click: parent.copyIRSQLToClipboard" data-clipboard-target="#irSQL">
+      <i class="fa fa-clipboard" aria-hidden="true"></i> Copy To Clipboard</button>
+  </div>
+  <div id='irSQL'>
+    <tabs params="
+        tabs: [
```
Comment (wivern) [inline js/pages/incidence-rates/tabs/ir-export.html]: Why not to point tabs in the same way as in the cohort definition page?

File: js/pages/incidence-rates/tabs/ir-export.js [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+	AutoBind,
+	commonUtils,
+) {
+	class IRexportTab extends AutoBind(Component) {
+		constructor(params) {
+      super(params);
+      this.parent = params.parent;
+      this.parent.showSql();
+    }
+	}
```
Comment (wivern) [inline js/pages/incidence-rates/tabs/ir-export.js]: tabs vs spaces

File: js/pages/incidence-rates/tabs/ir-generation.js [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+	Component,
+	AutoBind,
+	commonUtils,
+) {
+	class IRgenerationTab extends AutoBind(Component) {
+		constructor(params) {
+      super(params);
+      this.parent = params.parent;
+    }
+	}
```
Comment (wivern) [inline js/pages/incidence-rates/tabs/ir-generation.js]: tabs vs spaces

File: js/pages/incidence-rates/tabs/ir-utilities.js [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+	Component,
+	AutoBind,
+	commonUtils,
+) {
+	class IRutilitiesTab extends AutoBind(Component) {
+		constructor(params) {
+      super(params);
+      this.parent = params.parent;
+    }
+	}
```
Comment (wivern) [inline js/pages/incidence-rates/tabs/ir-utilities.js]: tabs vs spaces

File: js/services/IRAnalysis.js [hunk 16]
```diff
@@ -150,6 +150,27 @@ define(function (require, exports) {
...
+	function getSql(expression, analysisId, options) {
+		const data = {
+			expression: expression,
+			analysisId: analysisId,
+			options: options,
+		  };
+		const getSqlPromise = httpService.doPost(config.webAPIRoot + 'ir/sql', data);
+		getSqlPromise.catch((error) => {
+			console.log("Error: " + error);
+			authApi.handleAccessDenied(error);
```
Comment (wivern) [inline js/services/IRAnalysis.js]: What is the logic here? Why access denied handled here but not handled in the `translateSql` below?

File: js/pages/cohort-definitions/cohort-definition-manager.js (hunk 6 above)
Comment (ssuvorov-fls) [reply js/pages/cohort-definitions/cohort-definition-manager.js]: it is used in component export-sql.js

File: js/pages/incidence-rates/ir-manager.js (hunk 9 above)
Comment (ssuvorov-fls) [reply js/pages/incidence-rates/ir-manager.js]: it is used in component export-sql.js

File: js/services/IRAnalysis.js (hunk 16 above)
Comment (ssuvorov-fls) [reply js/services/IRAnalysis.js]: access denied handler was added to translateSql

File: js/pages/incidence-rates/tabs/ir-export.html (hunk 12 above)
Comment (ssuvorov-fls) [reply js/pages/incidence-rates/tabs/ir-export.html]: tabs/ir-export.* were removed from sources

File: js/pages/cohort-definitions/cohort-definition-manager.js [hunk 17]
```diff
@@ -677,6 +668,16 @@ define(['jquery', 'knockout', 'text!./cohort-definition-manager.html',
 				reportSourceKey: this.reportSourceKey,
 
 			}
+			this.translateParams = [];
+			globalConstants.dialects.forEach(d => {
```
[the remark this answers is not in the document]
Comment (ssuvorov-fls) [reply js/pages/cohort-definitions/cohort-definition-manager.js]: done

File: js/pages/incidence-rates/ir-manager.js (hunk 7 above)
Comment (ssuvorov-fls) [reply js/pages/incidence-rates/ir-manager.js]: done

File: js/pages/incidence-rates/tabs/ir-conceptsets.js (hunk 10 above)
Comment (ssuvorov-fls) [reply js/pages/incidence-rates/tabs/ir-conceptsets.js]: done

File: js/pages/incidence-rates/tabs/ir-definition.js (hunk 11 above)
Comment (ssuvorov-fls) [reply js/pages/incidence-rates/tabs/ir-definition.js]: done

File: js/pages/incidence-rates/tabs/ir-export.js (hunk 13 above)
Comment (ssuvorov-fls) [reply js/pages/incidence-rates/tabs/ir-export.js]: done

File: js/pages/incidence-rates/tabs/ir-generation.js (hunk 14 above)
Comment (ssuvorov-fls) [reply js/pages/incidence-rates/tabs/ir-generation.js]: done

File: js/pages/incidence-rates/tabs/ir-utilities.js (hunk 15 above)
Comment (ssuvorov-fls) [reply js/pages/incidence-rates/tabs/ir-utilities.js]: done

**Comment by anthonysena** [comment]
Closed in favor of the approach done in #2089

**Pull request closed without merging**